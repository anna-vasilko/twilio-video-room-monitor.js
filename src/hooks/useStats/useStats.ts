import React from 'react';
import { StatsReport } from 'twilio-video';
import { RoomStatsContext } from '../../components/RoomStatsProvider/RoomStatsProvider';

export default function useStats() {
  const stats = React.useContext(RoomStatsContext);
  return stats;
}

function getAllTracks(statsReport: StatsReport) {
  const { localAudioTrackStats, localVideoTrackStats, remoteAudioTrackStats, remoteVideoTrackStats } = statsReport;

  const allTracks = [
    ...localAudioTrackStats,
    ...localVideoTrackStats,
    ...remoteAudioTrackStats,
    ...remoteVideoTrackStats,
  ];

  return allTracks;
}

export function getTrackData(trackSid: string, statsReport: StatsReport) {
  const allCurrentTracks = getAllTracks(statsReport);
  return allCurrentTracks.filter((t) => t.trackSid === trackSid);
}

const sum = (arr: number[]) => arr.reduce((a, b) => a + b, 0);
const round = (num: number) => Math.round((num + Number.EPSILON) * 10) / 10;

export function useTrackBandwidth(trackSid: string) {
  const { stats, previousStats } = useStats();
  if (!stats || !previousStats) return null;

  const currentTrackData = getTrackData(trackSid, stats.statsReports[0]);
  const previousTrackData = getTrackData(trackSid, previousStats.statsReports[0]);

  if (
    currentTrackData.length === 0 ||
    previousTrackData.length === 0 ||
    currentTrackData.length !== previousTrackData.length
  ) {
    return null;
  }

  const currentBytes = sum(currentTrackData.map((d) => d.bytesReceived ?? d.bytesSent ?? 0));
  const previousBytes = sum(previousTrackData.map((d) => d.bytesReceived ?? d.bytesSent ?? 0));

  const currentTime = currentTrackData[0]?.timestamp;
  const previousTime = previousTrackData[0]?.timestamp;

  return round((currentBytes - previousBytes) / (currentTime / previousTime) / 1000);
}

export function useTotalBandwidth(kind: 'bytesSent' | 'bytesReceived') {
  const { stats, previousStats } = useStats();
  if (!stats || !previousStats) return null;

  const currentTrackData = getAllTracks(stats.statsReports[0]);
  const previousTrackData = getAllTracks(previousStats.statsReports[0]);

  if (currentTrackData.length === 0 || previousTrackData.length === 0) return null;

  const currentBytes = sum(currentTrackData.map((d) => d[kind] ?? 0));
  const previousBytes = sum(previousTrackData.map((d) => d[kind] ?? 0));

  const currentTime = currentTrackData[0]?.timestamp;
  const previousTime = previousTrackData[0]?.timestamp;

  return round((currentBytes - previousBytes) / (currentTime / previousTime) / 1000);
}
