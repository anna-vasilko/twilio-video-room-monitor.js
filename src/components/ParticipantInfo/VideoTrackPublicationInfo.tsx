import React from 'react';
import {
  LocalVideoTrack,
  LocalVideoTrackPublication,
  LocalVideoTrackStats,
  RemoteVideoTrack,
  RemoteVideoTrackPublication,
  RemoteVideoTrackStats,
  VideoTrack,
} from 'twilio-video';
import useIsTrackEnabled from '../../hooks/useIsTrackEnabled/useIsTrackEnabled';
import useIsTrackSwitchedOff from '../../hooks/useIsTrackSwitchedOff/useIsTrackSwitchedOff';
import useMediaStreamTrack from '../../hooks/useMediaStreamTrack/useMediaStreamTrack';
import usePublishPriority from '../../hooks/usePublishPriority/usePublishPriority';
import { useTrackBandwidth, useTrackData } from '../../hooks/useStats/useStatsUtils';
import useTrack from '../../hooks/useTrack/useTrack';
import useVideoTrackDimensions from '../../hooks/useVideoTrackDimensions/useVideoTrackDimensions';
import Datum from '../typography/Datum/Datum';
import MediaStreamTrackInfo from '../typography/MediaStreamTrackInfo/MediaStreamTrackInfo';
import StatsContainer from '../typography/StatsContainer/StatsContainer';

const getDimensionString = (dimensions?: VideoTrack.Dimensions) =>
  dimensions ? `${dimensions.width} x ${dimensions.height}` : 'undefined';

export const VideoTrackInfo: React.FC<{
  track: LocalVideoTrack | RemoteVideoTrack;
  trackSid: string; // Passing trackSid from the publication object because it not on the LocalVideoTrack object
}> = ({ track, trackSid }) => {
  const dimensions = useVideoTrackDimensions(track);
  const isSwitchedOff = useIsTrackSwitchedOff(track);
  const isEnabled = useIsTrackEnabled(track);
  const trackBandwidth = useTrackBandwidth(trackSid);
  const trackData = useTrackData(trackSid) as LocalVideoTrackStats | RemoteVideoTrackStats | null;
  const mediaStreamTrack = useMediaStreamTrack(track);

  let lossPercentage: string | null;

  if (trackData) {
    const { packetsReceived, packetsSent, packetsLost } = trackData;

    const totalPackets = packetsReceived ?? packetsSent;
    const numPacketsLost = packetsLost ?? 0;

    lossPercentage = totalPackets ? ((numPacketsLost / totalPackets) * 100).toLocaleString() : null;
  }

  return (
    <>
      <Datum label="Dimensions" value={getDimensionString(dimensions)} />
      <Datum label="isSwitchedOff" value={String(isSwitchedOff)} />
      <Datum label="isEnabled" value={String(isEnabled)} />
      <Datum label="Bandwidth" value={String(trackBandwidth?.toLocaleString()) + 'kbps'} />
      {trackData && (
        <>
          <Datum label="Codec" value={String(trackData.codec)} />
          <Datum label="Framerate" value={String(trackData?.frameRate)} />
          <Datum label="Packets Lost" value={String(trackData?.packetsLost)} />
          <Datum label="Packet Loss Percentage" value={String(lossPercentage!) + '%'} />
        </>
      )}
      <MediaStreamTrackInfo track={mediaStreamTrack} />
    </>
  );
};

export const VideoTrackPublicationInfo: React.FC<{
  publication: LocalVideoTrackPublication | RemoteVideoTrackPublication;
}> = ({ publication }) => {
  const publishPriority = usePublishPriority(publication);
  const track = useTrack(publication) as LocalVideoTrack | RemoteVideoTrack | undefined;

  return (
    <StatsContainer>
      <Datum label="Name" value={publication.trackName} />
      <Datum label="SID" value={publication.trackSid} />
      <Datum label="isSubscribed" value={String(!!track)} />
      <Datum label="publishPriority" value={String(publishPriority)} />
      {track && <VideoTrackInfo track={track} trackSid={publication.trackSid} />}
    </StatsContainer>
  );
};
