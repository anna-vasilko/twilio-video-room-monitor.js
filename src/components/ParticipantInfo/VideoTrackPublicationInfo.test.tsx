import React from 'react';
import { shallow } from 'enzyme';
import { VideoTrackPublicationInfo, VideoTrackInfo } from './VideoTrackPublicationInfo';
import { useTrackBandwidth, useTrackData } from '../../hooks/useStats/useStatsUtils';
import useTrack from '../../hooks/useTrack/useTrack';
import { RemoteVideoTrack } from 'twilio-video';

jest.mock('../../hooks/useIsTrackEnabled/useIsTrackEnabled', () => () => true);
jest.mock('../../hooks/useIsTrackSwitchedOff/useIsTrackSwitchedOff', () => () => false);
jest.mock('../../hooks/usePublishPriority/usePublishPriority', () => () => 'standard');
jest.mock('../../hooks/useStats/useStatsUtils');
jest.mock('../../hooks/useTrack/useTrack');
jest.mock('../../hooks/useVideoTrackDimensions/useVideoTrackDimensions', () => () => ({ width: 1280, height: 720 }));

const mockUseTrackBandwidth = useTrackBandwidth as jest.Mock<any>;
const mockUseTrackData = useTrackData as jest.Mock<any>;
const mockUseTrack = useTrack as jest.Mock<any>;

mockUseTrackBandwidth.mockImplementation(() => 1234.56);
mockUseTrackData.mockImplementation(() => ({ codec: 'testCodec', frameRate: null, packetsLost: 7 }));
mockUseTrack.mockImplementation(() => 'testTrack');

describe('the VideoTrackInfo component', () => {
  it('should render correctly if a video track is present', () => {
    const wrapper = shallow(<VideoTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly if there is no video track data', () => {
    mockUseTrackData.mockImplementationOnce(() => null);
    const wrapper = shallow(<VideoTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper.find({ label: 'Codec' }).exists()).toBeFalsy();
  });

  it('should render correctly for the RemoteVideoTrack when packetsLost is not defined', () => {
    mockUseTrackData.mockImplementationOnce(() => (
      { codec: 'testCodec', frameRate: null, packetsLost: null}
    ));
    const wrapper = shallow(<VideoTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly for the RemoteVideoTrack when both packetsReceived and packetsLost are defined', () => {
    mockUseTrackData.mockImplementationOnce(() => (
      { codec: 'testCodec', frameRate: null, packetsLost: 7, packetsReceived: 100 }
    ));
    const wrapper = shallow(<VideoTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly for the LocalVideoTrack when both packetsSent and packetsLost are defined', () => {
    mockUseTrackData.mockImplementationOnce(() => (
      { codec: 'testCodec', frameRate: null, packetsLost: 56, packetsSent: 183 }
    ));
    const wrapper = shallow(<VideoTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly for the RemoteVideoTrack when packetsLost is 0 and packetsReceived is defined', () => {
    mockUseTrackData.mockImplementationOnce(() => (
      { codec: 'testCodec', frameRate: null, packetsLost: 0, packetsReceived: 10 }
    ));
    const wrapper = shallow(<VideoTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly for the LocalVideoTrack when packetsLost is 0 and packetsSent is defined', () => {
    mockUseTrackData.mockImplementationOnce(() => (
      { codec: 'testCodec', frameRate: null, packetsLost: 0, packetsSent: 10 }
    ));
    const wrapper = shallow(<VideoTrackInfo track={{} as any} trackSid="testSid" />);
    expect(wrapper).toMatchSnapshot();
  });

});

describe('the VideoTrackPublicationInfo component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <VideoTrackPublicationInfo publication={{ trackName: 'testName', trackSid: 'testSid' } as any} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render the VideoTrackInfo component if a video track is not present', () => {
    mockUseTrack.mockImplementationOnce(() => null);
    const wrapper = shallow(
      <VideoTrackPublicationInfo publication={{ trackName: 'testName', trackSid: 'testSid' } as any} />
    );
    expect(wrapper.find(VideoTrackInfo).exists()).toBeFalsy();
  });
});
