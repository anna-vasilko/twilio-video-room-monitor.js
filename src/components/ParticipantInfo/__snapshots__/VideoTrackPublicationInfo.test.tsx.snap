// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the VideoTrackInfo component should render correctly for the LocalVideoTrack when both packetsSent and packetsLost are defined 1`] = `
<Fragment>
  <Memo(Datum)
    label="Dimensions"
    value="1280 x 720"
  />
  <Memo(Datum)
    label="isSwitchedOff"
    value="false"
  />
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Framerate"
    value="null"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="56"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="30.601%"
  />
</Fragment>
`;

exports[`the VideoTrackInfo component should render correctly for the LocalVideoTrack when packetsLost is 0 and packetsSent is defined 1`] = `
<Fragment>
  <Memo(Datum)
    label="Dimensions"
    value="1280 x 720"
  />
  <Memo(Datum)
    label="isSwitchedOff"
    value="false"
  />
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Framerate"
    value="null"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="0"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="0%"
  />
</Fragment>
`;

exports[`the VideoTrackInfo component should render correctly for the RemoteVideoTrack when both packetsReceived and packetsLost are defined 1`] = `
<Fragment>
  <Memo(Datum)
    label="Dimensions"
    value="1280 x 720"
  />
  <Memo(Datum)
    label="isSwitchedOff"
    value="false"
  />
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Framerate"
    value="null"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="7"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="7%"
  />
</Fragment>
`;

exports[`the VideoTrackInfo component should render correctly for the RemoteVideoTrack when packetsLost is 0 and packetsReceived is defined 1`] = `
<Fragment>
  <Memo(Datum)
    label="Dimensions"
    value="1280 x 720"
  />
  <Memo(Datum)
    label="isSwitchedOff"
    value="false"
  />
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Framerate"
    value="null"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="0"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="0%"
  />
</Fragment>
`;

exports[`the VideoTrackInfo component should render correctly for the RemoteVideoTrack when packetsLost is not defined 1`] = `
<Fragment>
  <Memo(Datum)
    label="Dimensions"
    value="1280 x 720"
  />
  <Memo(Datum)
    label="isSwitchedOff"
    value="false"
  />
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Framerate"
    value="null"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="null"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="undefined%"
  />
</Fragment>
`;

exports[`the VideoTrackInfo component should render correctly if a video track is present 1`] = `
<Fragment>
  <Memo(Datum)
    label="Dimensions"
    value="1280 x 720"
  />
  <Memo(Datum)
    label="isSwitchedOff"
    value="false"
  />
  <Memo(Datum)
    label="isEnabled"
    value="true"
  />
  <Memo(Datum)
    label="Bandwidth"
    value="1,234.56kbps"
  />
  <Memo(Datum)
    label="Codec"
    value="testCodec"
  />
  <Memo(Datum)
    label="Framerate"
    value="null"
  />
  <Memo(Datum)
    label="Packets Lost"
    value="7"
  />
  <Memo(Datum)
    label="Packet Loss Percentage"
    value="undefined%"
  />
</Fragment>
`;

exports[`the VideoTrackPublicationInfo component should render correctly 1`] = `
<styled.div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Memo(Datum)
    label="Name"
    value="testName"
  />
  <Memo(Datum)
    label="SID"
    value="testSid"
  />
  <Memo(Datum)
    label="isSubscribed"
    value="true"
  />
  <Memo(Datum)
    label="publishPriority"
    value="standard"
  />
  <VideoTrackInfo
    track="testTrack"
    trackSid="testSid"
  />
</styled.div>
`;
